scico/linop/xray/_xray.py

```diff
@@ -115,124 +94,362 @@ def __init__(
+
+    .. math::
+        \mathbf{M} \begin{bmatrix}
+        i + \frac{1}{2} \\ j + \frac{1}{2} \\ k + \frac{1}{2}
+        \end{bmatrix} + \mathbf{t} \,.
+
+    The detector pixel at index `(i, j)` covers detector coordinates
+    :math:`[i+1) \times [j+1)`.
+
+    :meth:`XRayTransform3D.matrices_from_euler_angles` can help to
+    make these geometry arrays.
 
-    return HTy
 
 
-@partial(jax.jit, static_argnames=["nx"])
-@partial(jax.vmap, in_axes=(None, None, None, 0, None))
-def _calc_weights(x0, dx, nx, angle, y0):
-    """
 
-    Args:
-        x0: Location of the corner of the pixel im[0,0].
-        dx: Pixel side length in x- and y-direction. Units are such
-            that the detector bins have length 1.0.
-        nx: Input image shape.
-        angle: (num_angles,) array of angles in radians. Pixels are
-            projected onto units vectors pointing in these directions.
-            (This argument is `vmap`ed.)
-        y0: Location of the edge of the first detector bin.
     """
-    u = [jnp.cos(angle), jnp.sin(angle)]
-    Px0 = x0[0] * u[0] + x0[1] * u[1] - y0
-    Pdx = [dx[0] * u[0], dx[1] * u[1]]
-    Pxmin = jnp.min(jnp.array([Px0, Px0 + Pdx[0], Px0 + Pdx[1], Px0 + Pdx[0] + Pdx[1]]))
-
-    Px = (
-        Pxmin
-        + Pdx[0] * jnp.arange(nx[0]).reshape(-1, 1)
-        + Pdx[1] * jnp.arange(nx[1]).reshape(1, -1)
-    )
-
-    # detector bin inds
-    inds = jnp.floor(Px).astype(int)
-
-    # weights
-    Pdx = jnp.array(u) * jnp.array(dx)
-    diag1 = jnp.abs(Pdx[0] + Pdx[1])
-    diag2 = jnp.abs(Pdx[0] - Pdx[1])
-    w = jnp.max(jnp.array([diag1, diag2]))
-    f = jnp.min(jnp.array([diag1, diag2]))
-
-    width = (w + f) / 2
-    distance_to_next = 1 - (Px - inds)  # always in (0, 1]
-    weights = jnp.minimum(distance_to_next, width) / width
-
-    return inds, weights
+
+    def __init__(
+        self,
+        input_shape: Shape,
+        matrices: ArrayLike,
+        det_shape: Shape,
+    ):
+        r"""
+        Args:
+            input_shape: Shape of input image.
+            matrices: (num_views, 2, 4) array of homogeneous projection matrices.
+            det_shape: Shape of detector.
+        """
+
+        self.input_shape = input_shape
+        self.matrices = matrices
+        self.det_shape = det_shape
+        self.output_shape = (len(matrices), *det_shape)
+
+        super().__init__(
+            input_shape=self.input_shape,
+            output_shape=self.output_shape,
+            eval_fn=self.project,
+            adj_fn=self.back_project,
+        )
+
+    def project(self, im):
+        """Compute X-ray projection."""
+        return XRayTransform3D._project(im, self.matrices, self.det_shape)
+
+    def back_project(self, proj):
+        """Compute X-ray back projection"""
+        return XRayTransform3D._back_project(proj, self.matrices, self.input_shape)
+
+    @staticmethod
+    def _project(im: ArrayLike, matrices: ArrayLike, det_shape: Shape) -> ArrayLike:
```

Comment:
`ArrayLike` seems unnecessarily broad for the return value typing: from inspection of the code, it would seem to always be a jax Array. The same comments applies to a number of other methods below.